### ISSUE TITLE
aria-checked on option and treeitem

### ISSUE THREAD

**Issue by smhigley** [w3c/aria#1052]
The option role includes `aria-checked` in supported states and properties, which inherits into treeitem. 

Since `aria-selected` exists on both and already indicates selected state, it's unclear what additional meaning `aria-checked` could convey. E.g. is there a use case for either of these currently valid option states?:

```
<div role="option" aria-selected="true" aria-checked="false">Selected, but not checked</div>
<div role="option" aria-selected="false" aria-checked="true">Unselected, but checked</div>
```

If not, I believe it should be removed from `option` (and `treeitem`). I've already run across the above markup (odd as it seems) in a product. It resulted from some confusion reading the ARIA spec and thinking `aria-selected` indicated current focus and `aria-checked` indicated selection, so updating to a more strict set of allowed attributes seems beneficial.

**Comment by scottaohara** [comment]
yeh.  seems to me `aria-checked` on an `option` should not be a supported property?  

i'm less familiar with `treeitem` allowances but do think it's weird to have both.  

Interested if there's something I'm missing here though?

**Comment by accdc** [comment]
Hi, aria-checked is supported on the option and treeitem roles to support tri-state checkability using the "mixed" value, which doesn't exist using aria-selected. There are indeed cases where this is necessary, and it does map to equivalent roles within the accessibility APIs for both of these roles.
All the best,
Bryan

**Comment by smhigley** [comment]
I've run into the tri-state use case before, but I don't think that's a good reason to support both `aria-checked` and `aria-selected`. If anything, that seems to be a reason to remove `selected` in favor of `checked`.

If they're both supported on the same element, it seems to me that they would need to mean different things, and be able to be individually added to the option or treeitem independent of each other. If we go that route, there should probably be some definition of what it means if an option or tree item is selected but not checked, and vice versa.

**Comment by scottaohara** [comment]
agree with Sarah here.  it's confusing that both are allowed, and anecdotally can say it's resulted in some issues I've seen where developers have argued they needed both.  

I can think of situations where this might be necessary for tree items, so again, don't have as much of an opinion there.  But less so for options where a more straight forward pattern of checkboxes / a single tri-state checkbox to indicate a mixed state wouldn't have been preferable.  But, I'm interested in learning about patterns that would require a deviation.

**Comment by JAWS-test** [comment]
>  aria-checked: ... or represents a group of other elements that have a mixture of checked and unchecked values (mixed). https://www.w3.org/TR/wai-aria-1.1/#aria-checked

* I would remove aria-checked from option, because in my understanding no entry in a select list displays a checked status for child entries.
* For the same reason I would leave aria-checked at treeitem, because in a tree structure exactly this is possible. aria-checked should only be used if parent entries show the status of child entries. Otherwise aria-selected should be used. This is quite confusing, but I think the danger is that you suddenly invalidate currently valid implementations if you change that.

**Comment by accdc** [comment]
You can't remove aria-checked from option without breaking current implementations like the following one.
http://whatsock.com/tsg/Coding%20Arena/ARIA%20Comboboxes/ARIA%20Comboboxes%20(Native%20Inputs,%20Multiselect%20Editable%20with%20Substring%20Match)/demo.htm

In this case, after typing a string like "hum" then pressing the down arrow to enter the listbox, the spacebar is required to toggle checkability. Implementing this using aria-selected doesn't work, I've tried this.

Option checkability is also used to indicate when all controls in a subsequent section are checked or selected, set to "true" if all are, "false" if none are, or "mixed" is some of them are. Whe have enterprise clients using this within complex UIs right now.

If you pull this out, it will break all of these.

**Comment by smhigley** [comment]
In general I think it makes sense to prioritize creating a clear and logical spec in the long term over maintaining support for all existing implementations, but in order to be considerate to existing implementations we could add a `deprecated` note and reserve breaking changes for major version changes.

Specifically regarding the use case you linked to: I'm not sure I understand why that's an argument for `aria-checked` over `aria-selected` on `role="option"`. It seems like a case where `aria-selected` actually has better support. Using that link, I get some unexpected screen reader output: in JAWS/Firefox, I hear "not selected, checked", VoiceOver on iOS doesn't seem to read the checked state at all, and Narrator seems to have some unrelated issues that make it difficult to test.

I put together a quick example with `aria-selected` ([https://www-ukpsneijli.now.sh](https://www-ukpsneijli.now.sh)), and I seem to be hearing the selected state correctly in JAWS and NVDA on FF and Chrome, Narrator on Edge/Edge Chromium, and iOS/VoiceOver. The only potential quirk is that both JAWS and NVDA read when an option is not selected and do not explicitly say "selected", but this matches behavior on native `<select multiple>` elements, so I'd consider that expected.

The mixed selection state does seem compelling to me for tree items, but I still think we shouldn't allow both attributes to coexist. If we do find a compelling reason to support both, we likely would need to define how both would map if they were independently defined on the same node.

**Comment by scottaohara** [comment]
yeh, can confirm the double state announcement quirk with Firefox+JAWS, as well as no selected items announced, regardless of how many are checked, with VO+Safari with the whatsock example.

Sarah largely summed up everything else I'd want to say here.  Regarding noting something as deprecated, [seems there's precedent.](https://www.w3.org/TR/wai-aria-1.1/#aria-dropeffect)

**Comment by JAWS-test** [comment]
@accdc: For the example http://whatsock.com/tsg/Coding%20Arena/ARIA%20Comboboxes/ARIA%20Comboboxes%20(Native%20Inputs,%20Multiselect%20Editable%20with%20Substring%20Match)/demo.htm your explanation that aria-checked must be used because of the mixed status does not apply. Therefore I would also say that aria-checked should not be applicable at least for option, but possibly for treeitem

**Comment by mcking65** [comment]
In general, we try to avoid putting limitations into the ARIA spec unless they are necessary. This is because we cannot anticipate every possible use case.

There are definitely use cases for both listbox and tree where checked is a more appropriate semantic than selected and vice versa. Ruling out either semantic is not really a necessary limitation.

I think there might be a potential argument for not allowing both at the same time in the same widget. However, is that something that should be warned against or prohibitted? Is it really the case that there is not a scenario where both could be useful?

Imagine a scenario where the checked state represents a persistent state that gets stored. And, in the same list or tree, there may be operations that can be performed on the items, e.g., remove, move higher in the list, etc. The selected state could represent selection for operation while the checked state represents the persistent state of the item. Maybe the instructions say check all items that you want stored offline. Additionally, you could manipulate the order of items. So, you could use standard selection keys to select one or more items for an operation like move up.

I am not convinced that putting a prohibition in the ARIA spec is appropriate in this case. Perhaps a middle ground approach is to consider an "authors "MAY" statement that says they "May" use both but "SHOULD" be careful to ensure users are able to discern the meaning of each state.

We can also add clarifying information to the guidance for aria-checked and aria-selected in the APG.

**Comment by scottaohara** [comment]
In HTML `option` elements do not allow the checked attribute.  In regards to parity with the host language, I do not understand why `aria-checked` would be allowed (in HTML) as it would create a deviation from native behavior, and allows for the introduction of author and potentially user confusion. 

Please don't take this as an argument to me not understanding there being a potential use case to indicate that some / all / none of the options within a listbox have been selected.  I just question if _this_ is the way to do it, again noting that even the provided example doesn't expose this information consistently.  

That said, if such allowances for both states continue then I would fully agree with your mentioning of a middle round, as I do think the specification should document why this diverges from HTML, and guidance of when both states would be appropriate could then make it's way into APG.

**Comment by accdc** [comment]
Hi,
Regarding role parity, there is a checkable listbox windows class, which is what aria-checked already ties into on the Windows OS going back many years.
https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.checkedlistbox?view=netframework-4.8

This is true in UIA as well, and can be seen every time you open My Documents. If you look at the accessibility tree properties for every item listed there, you will se a checkable state option on each option listed.

**Comment by smhigley** [comment]
@accdc it's an interesting example, but I'm not sure WinForms is the best framework to use as a blueprint going forward. Here's the same control in UWP, where you can see selection is handled with the `SelectionPattern`, and not the `TogglePattern`: https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/listview-and-gridview#item-selection-and-interaction. For context, WinForms (or Windows Forms) is an older framework and its UIA support is less robust than the newer WPF or UWP. It might also be worth noting that `aria-checked` is not currently mapped to anything in UIA or AX API for either `option` or `listbox` in Edge or Safari.

Regarding Matt's suggestion that we continue allowing both checked and selected:

I'm still concerned about allowing both `aria-selected` and `aria-checked` on the same element. For one, how would a user toggle each property separately? Even if there were a custom keyboard shortcut that could be communicated to screen reader and keyboard users, would it be possible to support switch devices or touch-based screen readers?

I've actually run into real-world use cases where tree items were both selectable and supported actions like move, delete, edit, etc., but none of those were implemented using dual checked/selected states. I don't think that's a good use case for needing both properties supported at once, since more usable implementations already exist.

I feel like maybe I could use some clarification on what would make a limitation necessary vs. not necessary. From my perspective, the risk of developer confusion and misuse combined with a lack of practical support are a compelling reason to limit this in spec, though I'm not sure how that matches up to a bar of necessary. Should we also be applying that bar to revisiting whether e.g. prohibiting dual checked/selected states on `tab` or a treegrid `row` is necessary?

**Comment by LJWatson** [comment]
Part of the problem seems to be that both attributes can be used without constraint. Enabling something to be simultaneously selected and unchecked (or vice versa) seems like something we should avoid.

Unless we can find a reasonable number of production examples in the wild, where there is a clear case for needing both attributes simultaneously, I suggest we make it a clear requirement that only one of these attributes can be used at a time.

**Comment by smhigley** [comment]
Just as a bit of a reality check, this is code I ran across today in a popular third-party UI framework that makes grand claims about accessibility (ids, classes, and text modified or removed so as to not call anyone out):

```
<li role="treeitem" aria-checked="true" aria-selected="false" aria-expanded="false">
  <div>
    <span role="presentation">
      <input type="checkbox" tabindex="-1">
      <label role="checkbox" aria-checked="true"></label>
    </span>
    <span>tree item text</span>
  </div>
</li>
```

Obviously they haven't implemented the treeitem well, and on some level we can't control that. On the other hand, this is actually a pretty tame misuse of ARIA compared to what I come across daily. When I'm able to ask the original developer why those choices were made, it's often "I saw this was supported in the ARIA spec" or "I saw this in the APG."

I know it's not our job to prevent developers from abusing ARIA or using APG examples without testing support, but I would really love to see the spec move to where a developer who has no specialist knowledge or context making a good-faith effort can take it at face value and end up with something reasonably accessible.

In this specific case, I think adding a requirement that `aria-checked` and `aria-selected` cannot both be present within the same `tree`, and removing `aria-checked` from `option` would aid in that without limiting any known or likely future valid use cases.

**Comment by craigkovatch** [comment]
> Imagine a scenario where the checked state represents a persistent state that gets stored. And, in the same list or tree, there may be operations that can be performed on the items, e.g., remove, move higher in the list, etc. The selected state could represent selection for operation while the checked state represents the persistent state of the item. Maybe the instructions say check all items that you want stored offline. Additionally, you could manipulate the order of items. So, you could use standard selection keys to select one or more items for an operation like move up.

Here is a real-world case in Tableau where we have a listbox with options that are both checkable and selectable, just as you've described:

```
  listbox name="Year" aria-multiselectable={true}
    option name="(All)" aria-checked={false} aria-selected={false}
    option name="2014" aria=checked={true} aria-selected={true}
    option name="2015" aria=checked={true} aria-selected={true}
    option name="2016" aria=checked={true} aria-selected={false}
    option name="2017" aria=checked={false} aria-selected={false}
```
![image](https://user-images.githubusercontent.com/10970257/93542930-23e11280-f90f-11ea-9290-e022cd45e4be.png)

In this case, three of the five options are checked, which indicates the user wishes to filter on those three years. Two of the options are selected, which allows a user (via either mouse or keyboard) to batch-select or unselect a large number of items.

If there is another pattern that would be better for representing this, I'm happy to hear about it. But this seemed like the sanest of the ARIA primitives available.

**Comment by craigkovatch** [comment]
Speaking of "in the wild" examples, here's another: https://a11y-guidelines.orange.com/web_EN/exemples/check-listbox/index.html

**Comment by stes-acc** [comment]
> > Imagine a scenario where the checked state represents a persistent state that gets stored. And, in the 

>   listbox name="Year" aria-multiselectable={true}
>     option name="(All)" aria-checked={false} aria-selected={false}
>     option name="2014" aria=checked={true} aria-selected={true}
>     option name="2015" aria=checked={true} aria-selected={true}
>     option name="2016" aria=checked={true} aria-selected={false}
>     option name="2017" aria=checked={false} aria-selected={false}
> ```
> 
> ![image](https://user-images.githubusercontent.com/10970257/93542930-23e11280-f90f-11ea-9290-e022cd45e4be.png)

This is very common in Business UIs (Material selection scenarios etc,)

**Comment by smhigley** [comment]
Both the linked examples ([Orange](https://a11y-guidelines.orange.com/web_EN/exemples/check-listbox/index.html) and [Lightning](https://www.lightningdesignsystem.com/components/combobox/#Listbox-markup)) seem to be mixing up `aria-selected` with focus/`aria-activedescendant`. Using `aria-selected` to denote the current option is not correct ARIA usage, and is exactly the sort of misconception I think could be cleared up by disallowing both checked and selected at once. @stes-acc all the cases I've come across in common UI libraries fall into this bucket, so they're more of an argument for fixing this than leaving it as-is :).

@craigkovatch for your example, I am *very* interested in what keys a user would press if they were on e.g. the checked-but-not-selected "2016" option in your example if they wanted to a) uncheck, or b) select that option.

I'm also interested in how you communicate that separate keyboard interaction for checked vs. selected to the user, and also how you communicate the different in-app meanings of checked vs. selected. I don't even think this works visually -- as a sighted mouse user, if I'm able to separately check and select by e.g. clicking on the checkbox vs. the option name, I would have no idea which one I need to do in order to accomplish any given task. And then there's the added problem of awful pointer target sizes with a checkbox nested within a larger option hit target, and the problems go on...

I really think having multiple on/off states on any one item like this fails along multiple axes, and shouldn't be supported in the spec.

**Comment by carmacleod** [comment]
@smhigley I agree completely that Orange and Lightning (good name for a band) are mixing up selection and focus, and they actually don't need to use aria-selected. I tried to make this clearer in https://github.com/w3c/aria/pull/1326 - please take a look. (Of course, an edge case came up that muddied the clarity a bit, but I think the new sentence will at least be somewhat helpful.)

I believe that in @craigkovatch's example, down arrow will select "2016" (or shift+down arrow will select "2014, 2015, 2016"), and then space will uncheck it (or, if shift+down arrow was used, space will uncheck "2014, 2015, 2016"). Another space would check again.

I think the idea would be to be able to batch up the filtering. I could see this potentially making a difference in the performance of the filtering code (i.e. only filter when things become checked or unchecked), e.g. the user could carefully select 27 contiguous and discontiguous years and then hit space to filter them all at once - which may involve only one round trip to a server or something like that.

Having multi-selection may also make it a bit more efficient for users to do "batch checking" by selecting first with typical mouse and/or keyboard multiselect actions, and then checking/unchecking many items at once with the space key.

Of course, if the checkmarks are just visual and are actually semantically identical to selection, then aria-checked would not be needed and the user's multiselect actions would cause the checking and filtering, which would be _almost_ the same as having them as separate states... except for that performance thing.

I'm not saying that it isn't confusing to have selected and checked in the same container - it is. We should make sure the guidance is very clear about that, and ask authors to think carefully about their use case. We should also make sure to explain the concept that "focus does not equal selection". And we should tell authors to prefer aria-selected over aria-checked, even when their items have a checkmarky visual treatment. But I don't think we should make it a violation to have both.

@craigkovatch, please confirm my guess that down arrow selects and space unchecks, and please also let us know:
- is all of the keyboard behavior the same on Windows and Mac?
- is all of the mouse behavior the same on Windows and Mac?
- does Enter (aka Return) do anything on either platform?
- does mouse double-click do anything on either platform?

Thanks!

**Comment by craigkovatch** [comment]
@smhigley on this particular control, selection follows focus (not shown in my screenshot but indicated with a 1px blue outline). If focus was on 2015 and I press down arrow,14 2016 would be the only selected item, the focused item, and unchecked. Space then toggles the checked state. If focus was on 2015 and I press shift+down arrow, now 2014+2015+2016 are all selected and checked. Pressing space would clear the checked state from all of them.

I agree with you that there are numerous ways this control could be improved. Nevertheless, it *does* exist, it has existed and worked this way for 10+ years, and it is unlikely to change any time soon. If we simply say "sorry, spec doesn't support that", we aren't going to help anyone -- not developers, and certainly not AT users.

> as a sighted mouse user, if I'm able to separately check and select by e.g. clicking on the checkbox vs. the option name, I would have no idea which one I need to do in order to accomplish any given task. 

The fact is that hundreds of thousands of sighted mouse users use this control every day, and have somehow figured it out. So I think comments like largely fall into the territory of opining about how things should be in an ideal world. I largely agree with you about how the ideal world should look -- I give opinions like this to the designers at my company very frequently :)

But I think we make no progress for anyone if we only support an ideal world. So, with apologies for some preaching to the choir here: moving the state of accessibility requires a certain amount of meeting the world where-it-is, and that includes controls and interactions like this which (we agree) are non-ideal but nonetheless need a way to expose themselves to AT. @accdc's examples from proto-Windows show that this is very old prior art; we can't just pretend it doesn't exist and write a spec that will never support it.

**Comment by craigkovatch** [comment]
> I think the idea would be to be able to batch up the filtering.

@carmacleod yes exactly. And our research has shown that keeping users "in the flow" is extremely important for the kind of exploratory visual analysis that our product enables. If I have to wait 4s for a round trip (whether to the server, or to the visualization engine) every time I change a checkbox, I lose my flow state quickly.

To be clear: the checkboxes are the real inclusion/exclusion of items in the data filter. The gray selected state is internal to the control, i.e. for batching operations as discussed.

> is all of the keyboard behavior the same on Windows and Mac?
Yes, kind of. I work only with our web product, and so our interactions are mostly identical between Windows and Mac, except for e.g. supporting CMD+ instead of CTRL+ for certain keyboard shortcuts. (There may be differences on our native-executable "desktop" product, where this control was originally birthed, but that's of course out of scope for ARIA specs.)

> is all of the mouse behavior the same on Windows and Mac?
Yes

> does Enter (aka Return) do anything on either platform?
Not in the screenshot shown. There is a version of this control, rarely used, which also has "Apply/Cancel" buttons, for the scenario that I have potentially hundreds of items that I want to manipulate before the server round-trip you mentioned. When these buttons are shown, ENTER commits, i.e. same as clicking the Apply button.

> does mouse double-click do anything on either platform?
In terms of there being an `onDoubleClick` handler, no. Clicking the checkbox twice would of course cause it to "do a full 360" :)

**Comment by sinabahram** [comment]
This keyboard behavior is exactly what I would expect as a blind user and matches mine and others' descriptions on our various calls. I very much do not want it to be forbidden by the spec. That would be a huge step backwards. Thanks for sharing this example.

**Comment by smhigley** [comment]
@craigkovatch how do you select multiple non-continguous items if the spacebar is reserved for checking? How does this work with a touch-based screen reader? Or voice control? If the checkbox is the main action, do people who use touch or a pointer with less fine control really need to aim exactly for the checkbox to perform that main action? If all selection does is allow you to check multiple items at once for better performance, how would you convert a selection into a checked state without the keyboard?

Disallowing both checked and selected at the same time on the same control isn't just an opinion -- it comes from the basic impossibility of making that dual state and dual interaction accessible. Essentially, if you have a single non-editable interactive item (a button, a link, an option within a listbox, etc.), you're limited to one single action a user may perform on it if you want to make it accessible and work across platforms and input modalities (leaving aside context menus at the moment).

This is one reason why controls with both a primary action and a dropdown like comboboxes and splitbuttons are so hard to make accessible to touch, touch-based screen readers, and other pointer interactions. To make them work, you need to expose an additional button to allow everyone to both perform the primary action and also separately expand/collapse the dropdown. This case is similar: if you want to allow everyone to separately toggle two different states on your listbox option, then you need two separate controls.

**Comment by craigkovatch** [comment]
@smhigley Yup, as I said, I agree with you. This control falls over in pretty much all of those ways. But it is an excellent case study of the unfortunate reality is that for legacy software like this, we're always going to be retrofitting as much AT support as possible while also balancing other business priorities and resource constraints.

Nonetheless, we still need a way to expose what it *does* do to AT in a sane way. Not everything can be perfect. We cannot make perfect the enemy of good. The spec is not useful if it only applies to ideal software.

**Comment by smhigley** [comment]
@craigkovatch you can continue marking up your control in any way you want to; the ARIA spec isn't going to make your current control suddenly break, or even change how screen readers treat checked + selected when the two do occur in tandem.

We shouldn't constrain the ARIA spec to support legacy patterns that are fundamentally inaccessible; if there are no valid, accessible use cases, then it should not be supported. We used to support `aria-expanded` on `role="dialog"` for example, and removed it because an expandable `role="dialog"` would be inaccessible by design. Having both `aria-checked` and `aria-selected` on the same option is similarly inaccessible by design, for all the reasons I mentioned + some cognitive accessibility concerns.

@sinabahram I know we discussed this in more depth on the deep dive, and if I remember correctly using `aria-current` + point of regard + constraining options to only one of checked or selected addressed all the use cases raised. You would still be able to have the keyboard behavior you mentioned without needing to introduce the accessibility problems of dual checking/selecting.

**Comment by craigkovatch** [comment]
> or even change how screen readers treat checked + selected when the two do occur in tandem.

Surely that is not true in the long term -- otherwise what's the point of having a spec?

> We shouldn't constrain the ARIA spec to support legacy patterns that are fundamentally inaccessible; if there are no valid, accessible use cases, then it should not be supported.

I think we can agree that accessibility is a spectrum, right? That there are experiences that are "more" or "less" accessible than others. I am trying to present a use case that, yes, falls over in some scenarios that you illustrate, but still derives clear value out of the two properties, even occurring in tandem. (Again, happy to hear suggestions about a better way to structure this in the current world.)

I agree with you that this is not a best-practice pattern, not in the slightest. But if we change the ARIA spec to disallow it, it will be flagged by e.g. accessibility auditors as Wrong, and thus lead to a situation where we create a less-useful experience because of a (forgive me) artificial constraint in the spec.

Perhaps I am misunderstanding you, but what you're proposing sounds very rigid to me -- either it's All or Nothing. I feel there's middle ground to be found here, along the lines that @mcking65 and @carmacleod have suggested. If not, then I think we might need to discuss bigger changes here. If enough people have gotten this wrong already, then there's a lacking in the tools we have available that will not be solved by just deleting `aria-checked` from `option`.

**Comment by sinabahram** [comment]
@smhigley 

>you can continue marking up your control in any way you want to; the ARIA spec isn't going to make your current control suddenly break, or even change how screen readers treat checked + selected when the two do occur in tandem.

I don't agree with that. The screen readers will happily point to the ARIA spec in order to justify something being unsupported. So, you are of course technically correct that the spec won't immediately break anything, but it certainly can/does/will affect software choices in the future that will break things.

> We shouldn't constrain the ARIA spec to support legacy patterns that are fundamentally inaccessible; if there are no valid, accessible use cases, then it should not be supported.

May I please ask for the official definitions of both `valid` and `accessible` as used by the spec? These are both heavily overloaded terms. What exactly must they mean in the specific context of the ARIA spec? I feel like without precise shared definitions, that assertion will mean different things to different people.

> I know we discussed this in more depth on the deep dive, and if I remember correctly using aria-current + point of regard + constraining options to only one of checked or selected addressed all the use cases raised. You would still be able to have the keyboard behavior you mentioned without needing to introduce the accessibility problems of dual checking/selecting.

I'm confused because that is not my understanding. To help me understand, are you stating that with the above pattern, as you have highlighted it, I can use my arrow keys to navigate, shift+arrow keys to select, control+space to do noncontiguous selection, and hit space on a selection to check multiple items at once? I can with the above pattern, which is very similar if not identical to what I was talking about on the call in terms of use cases. My understanding, just to be clear, is that what you are suggesting will unequivocally not allow for what I just said. If I am wrong about this, please let me know.

**Comment by smhigley** [comment]
Perhaps I should clarify what I mean when I said this pattern isn't accessible: this pattern is fundamentally impossible to make operable by someone who doesn't have access to a keyboard. This includes the following scenarios and AT:
- Anyone on a touch-only device, including touch-based screen readers
- Pointer-emulating AT like eye gaze or switch + mouse grid
- Voice Control

It also causes additional problems in the following cases:
- Low vision/use of color: you now need to differentiate between for states -- rest, checked, selected, and active
- Limited dexterity: a mouse user is going to have trouble clicking the checkboxes without accidentally changing the selection instead
- Cognitive accessibility: introducing two separate on/off states on the same control also introduces a lot of guesswork and lack of clarity around what each of those states means and how to interact with them
- The shift/control + arrow key selection behavior is not well known in practice, particularly for non-contiguous selection

As far as I'm concerned this isn't "accessible on a spectrum". These are deal-breakers.

The extensive usability and accessibility problems with this pattern are the reason I don't see it anywhere in the real world. Other than Craig's screenshot, I have literally never come across a UI that made me separately check and select the same control. And based on our discussion, it seems clear that that example has all of the accessibility problems outlined above, so it seems to me like an accessibility validator would be correct to flag it. There's even a very clear accessible alternative: add an "apply" button.

Again, this isn't some minor "less than ideal" quibble. The issue is that this pattern, by its very design, is actually *impossible* to make accessible to multiple common input methods. We also have yet to come across a use case that could not be solved in another way. In my opinion, patterns like that should not be supported by the ARIA spec.

**Comment by craigkovatch** [comment]
Thank you for the very clear thoughts @smhigley . Again I can’t help but say I understand and agree with where you’re coming from. But I am also faced with a lot of other real-world constraints that I feel you are ignoring.

To address some of your points: The control is already operable by the two most common input methods. It may not be 1-to-1 identical experience for both, but it does work. You bring up touch, but I’ve already said that there is a separate UI entirely presented for touch. You bring up  visual state differentiation — but it already does have four visual states (again, not all pictured). You say you’ve never seen this anywhere else, but there are several other linked examples provided by myself and others. You might hate them as UX patterns, and I wouldn’t argue, but saying they don’t exist is not true 🤷‍♂️ “I’ve never seen it” is not the same as “it doesn’t exist”.

Much more importantly I think the discussion misses the forest for the trees. I don’t think we should continue debating this particular control, because I am only intending it to be illustrative of a scenario which is going to be very common in extant, real-world software, and will apply to many Issues beyond this one.

Ignore for a second this particular funky listbox, because we’re talking about philosophy now. I am faced with the real-world constraint that this example — like many more in other pieces of software and companies — represents UI that exists, has existed for a long time and serves thousands/millions of users, and for some reason or another (be those good or bad, fair or unfair, inclusive or shitty reasons), will not be changed.

If that is the scenario, the spec change you are suggesting means *less* information is able to be provided to AT. Yes, there is a spectrum of accessibility. You can’t throw the baby out with the bathwater and say that, because you see issues with it (issues which you are quite right about), there is no value whatsoever in exposing *as much* as possible. Isn’t it better for it to mostly-work than mostly not-work?

There are always going to be tradeoffs made, especially by businesses attempting to improve legacy software. I believe very strongly that spec-writing of this kind should encourage and document best practices, but also attempt to provide a path for graceful degradation.

**Comment by smhigley** [comment]
@craigkovatch I work on practical accessibility problems for hundreds of products, including ones as complex as Tableau. Saying I'm ignoring real-world constraints is a bit... odd. Additionally, misrepresenting what I've written by saying I "hate" the pattern is unhelpful, untrue, and verges on sexist.

The reason I raised this issue and the reason I think it's important is because allowing both `aria-checked` and `aria-selected` on options and treeitems confuses developers in practice, who then mix up selection and focus/point of regard. The linked [Orange](https://a11y-guidelines.orange.com/web_EN/exemples/check-listbox/index.html) and [Lightning](https://www.lightningdesignsystem.com/components/combobox/#Listbox-markup) listboxes are excellent examples of exactly this problem in the real world.

If, as you say, there are "many more in other pieces of software and companies," could you provide links to some of those examples? Particularly ones that are public and that we can inspect and interact with. So far, yours is the only example to come out of this discussion that uses dual selected and checked states -- all the others only have one selected or checked state + point of regard.

**Comment by scottaohara** [comment]
Would agree with Sarah here.  I'm finding the responses to her to be rather unfortunate and diminishing of how much experience and knowledge she has with real world software here.  Please, be mindful of that and stop insinuating she is coming from a place of purely opinion.

I would personally like to see other working examples, as I also find the Orange and Lighting, as well as other linked demos thus far, lacking in accurately demonstrating what was otherwise outlined [in this comment](https://github.com/w3c/aria/issues/1052#issuecomment-694585197).  And as Sina summarized:
>I can use my arrow keys to navigate, shift+arrow keys to select, control+space to do noncontiguous selection, and hit space on a selection to check multiple items at once...

Unless I just don't understand what keys to use, or if they're somehow windows specific, I haven't been able to select more than a single item to check/uncheck at once.

**Comment by jnurthen** [comment]
@craigkovatch I'm really interested in how you present a different UI to users using touch on a Windows laptop.... I didn't think there was a way to do this.

**Comment by sinabahram** [comment]
@scottaohara  RE the keystrokes, I think I understand your question to be one of two possible things. Are you asking about selection, which is just hold down shift key and use arrow keys, or are you asking about noncontiguous selection, which is hold down control and then you can use space and arrow keys to select and navigate without selection. I wasn't referring to anything knew, just what's been true for the last 20 years or so of my computer usage. Please note that multiple screen reader users on our last depe dive call confirmed this expectation of keyboard usage. I'm not claiming people said it was easy or good for novices, but I did not hear any disagreement on the technical correctness of the above description of how continguous and noncontinguous selection occur. Please note that yes I'm referring to Windows specific keystrokes. @cookiecrook  confirmed, I believe, that noncontinguous selection with keyboard wasn't/isn't a thing on Mac. I just wish to humbly point out, that doesn't mean it shouldn't be a thing, full stop. @cookiecrook  if I have that wrong, my apologies, and maybe you can provide @scottaohara  the keystrokes for noncontinguous selection with keyboardthe .

It would be good to know where to discuss the off-topic spec stuff. I must admit to complete and utter confusion at this point about the rules under which we operate. I really don't understand the model we are operating under here, clearly. It feels to me like @smhigley  is applying the constraint that an interface only can count as an example if it itself is fully accessible, but I did not know this to be the operating constraint of the ARIA WG. Just on a call a week or two ago, it was stated that ARIA needs to make what exists out there accessible, but now I am getting the sense that ARIA should only try to facilitate fully accessible interfaces, even when lots of affordances may not rely upon ARIA what-so-ever. I'm not taking a position on the above. I sincerely desire to know what rules exist so I can contribute more effectively. To contextualize my confusion, I understand the argument to be as follows; we should not allow both selected and checked.because to do so implies an inaccessible interface. I simply wish to know if this constraint applies to every single affordance available in ARIA. If so, I need to calibrate my understanding.

@smhigley  if I have not understood the above clearly, please let me know so that I may better understand your position. I strongly believe we have the exact same goals here, which is to make the web better for everyone. As someone who benefits directly from that work, you only have my gratitude. I simply don't understand why we are wanting to forbid something that does exist and is helpful, even if it is for a small subset of users. That inaccessible interface is not going to be made accessible if the spec changes to forbid checked and selected being used together, but it is very likely that screen readers will have no impetus to support checked and selected together if we forbid it. This feels to me like a net-negative; hence why I keep pushing on it.

**Comment by craigkovatch** [comment]
@smhigley , I am not insinuating anything about your depth of experience. I am aware of it, and have told you many times how much I appreciate and respect it. I’ve reached out to you personally for advice on many occasions. But I feel your comments on this topic are too focused on “it has to be perfect or it can’t be saved”.  It was not my goal to discuss this specific control in depth, but to illustrate the practical ramifications of the problem it presents, and others like it, as we attempt to craft good specs as a community.

It is completely baseless to suggest the concerns I am raising are motivated by sexism. I’m not sure how the suggestion that you “hate” a pattern would signify that — based on posts and presentations of yours, I would say you “hate” tooltips as well, but that’s got nothing to do with anyone’s gender.

I am motivated by trying to find a solution that works for both best practice and not-best practice. I would still welcome your ideas on how this situation, and thus others like it, might be solved alternatively if your suggestion of removing aria-checked is adopted. I understand your opinion is this control should just not exist, but it does, and I still want to know how to do best for our users if this spec change is made.

**Comment by craigkovatch** [comment]
> @craigkovatch I'm really interested in how you present a different UI to users using touch on a Windows laptop.... I didn't think there was a way to do this.

Our support for desktop touch is poor. We assume if we see touch events that the user is on a mobile device and show different UI with larger tap targets and other touch accommodations in that case.

**Comment by craigkovatch** [comment]
@scottaohara I’m on board with everything you said. I don’t have examples outside my own scope, but it does seem to me there is other prior art here, eg the old WinForms screenshot that was offered.

I enthusiastically agree that we should probably figure out a discussion venue for “philosophy” of spec. I think we will have a hard time coming to a consensus on specific problems like this one without nailing down broader questions of approach like that.

And to echo @sinabahram (whose comments I also agree completely with) — we all have the same goal here. We all want the web to be better for everyone ❤️

**Comment by scottaohara** [comment]
@sinabahram i was specifically just trying to get provided examples to allow me to hit Shift under the assumption that up or down arrows would let me "select" additional options.  They did not.  I tried control instead and that just initiated the macOS function of showing all the open windows on my screens for re-ordering.  I then threw up my hands.

**Comment by joanmarie** [comment]
Some brief observations / opinions (and admittedly vast oversimplifications):
* If it can be selected (in the highlighted sense of that word), then it seems to me that `aria-selected` should be supported.
* If, independent of its selected state, it can be toggled (e.g. it has a visible checkbox or radiobutton), then it seems to me that `aria-checked` should be supported.
* Authors gonna author. Whether we support `aria-checked` on options/treeitems or not, some author is going to decide that their option/treeitem needs to have a checkbox. Screenshots of examples have been provided. So.... If we were to remove `aria-checked` support on these roles, the aforementioned author might decide to cram the checked state into the accessible name to make the screen reader do the author's bidding. This strikes me as making things worse; not better.
* If we really, really don't want to support `aria-checked` on options, then we probably should make options not children-presentational so authors can add checkable descendants. This constraint does not seem to be present on treeitem.
* If there is confusion about what `aria-checked` means versus what `aria-selected` means with respect to an option and/or tree item, maybe *that* is the thing to fix in the spec and address in the APG.
* If there is lack of support for the "checked" state in screen readers for the option and/or treeitem role, let's file bugs against the screen readers.

**Comment by craigkovatch** [comment]
> If we really, really don't want to support aria-checked on options, then we probably should make options not children-presentational so authors can add checkable descendants. This constraint does not seem to be present on treeitem.

Without speaking to how easy/hard/possible this would be, I think this is a GREAT idea to discuss. I have found no matter how many warnings and constraints I put in our library, “authors gonna author” and want to have interactive children in their options.

**Comment by scottaohara** [comment]
per Joanie's points, I would lean far more towards:
>If there is confusion about what aria-checked means versus what aria-selected means with respect to an option and/or tree item, maybe that is the thing to fix in the spec and address in the APG.


again, the feedback I've had on patterns like these, is that people don't understand the difference between checked and selected.  to non-spec-savy people (e.g., users and devs) these can mean the same thing.  hence, if we're going to continue to support things like this, we need to be clear to implementors and authors when/why/and how to use these states.  Also, I would personally love to see the future guidance indicate about how authors **don't** need to just reach for a pattern like this.  e.g., just because there's visually a checkmark, that doesn't always mean it _needs_ a checkbox, if that checkmark is just another way to visually indicate a selected state.



I would also want to lean as far away from the following as possible:
>If we really, really don't want to support aria-checked on options, then we probably should make options not children-presentational so authors can add checkable descendants.

**Comment by craigkovatch** [comment]
> I would also want to lean as far away from the following as possible

@scottaohara could you kindly elucidate your concerns on this? I expect there is much here that I "don't know I don't know", would love to understand why you feel strongly against it.

**Comment by scottaohara** [comment]
A number of concerns, really.

1. Blurring the lines of what interactive roles allow for nested interactive children more than there already is. Re: `treeitem`, an argument could be made that since there's no host language (HTML) equivalent then there is more leniency for allowances. That hypothetical argument falls apart for other roles which have HTML equivalents where an allowance like this would be in opposition to the requirements of it.
2. further distancing the allowances/expectations of `role=option` vs `<option>` (which again, does not support a checked state in native HTML, but at least neither ARIA or HTML are currently in misalignment regarding allowed children).  
3. The general usability of nested controls, e.g., the difficult click/touch areas, ensuring each control has an adequate accName without producing an overly redundant accName of either... etc.

i think that's enough for now.

**Comment by craigkovatch** [comment]
Thanks @scottaohara. #1/#2 are more persuasive to me than #3. Please see e.g. a whole bunch of Material UI examples of list items with nested/secondary controls: https://material-ui.com/components/lists/#list-controls -- more examples of "authors gonna author" and so I think we should try to determine what the best modeling is. They seem to be getting around this whole debate by using `<li>` rather than `<option>`, although that seems wrong in a different way to me (correct me if I'm wrong but li/role=listitem shouldn't be interactive).

Also, and I'm a hypocrite for saying so, but I think we should start at least one new thread as this Issue has several different lines of discussion to follow now. I'm not sure the right way/place(s) to do that, though.

**Comment by jscholes** [comment]
I think @joanmarie has summed up my thoughts in the most straight-forward, sensible way.  I'm mindful of the fact that the title of this issue is, "aria-checked on option and treeitem".  The overlap with other ARIA attributes is appropriate, but I feel has slightly derailed the discussion.

Speaking in practical terms, I'm interested in many use cases for having a checkbox-style look and feel within a list box and tree view where selection follows focus.  This has come up with multiple client projects, some of which are public and can be discussed here, some of which can't.

It is a fact that screen readers do facilitate traversal through options or tree items which, in such cases, wouldn't cause selection or focus to move, but would still allow the checked state to be queried.  As an example, object navigation in NVDA would permit such exploration inside both list boxes and trees.

As a practical example, the sidebar on this GitHub issue contains several "dropdowns" (for lack of a better term) for applying assignees, labels, etc.  Currently, expanding those sidebar sections reveals a search input and menu, which is a semantic hack to force a screen reader to recognise the checked semantics via `role="menuitemcheckbox"`.  But it poses several issues, and I'm sure many in this thread have concerns about the UX which aren't strictly relevant.  Here's what I'd love to see:

1. An editable combo box with list box pop-up, where the combo box is editable to search for labels, users, etc.
2. A "list autocomplete with automatic selection" pattern (in APG terms).
3. `aria-selected` being used to follow focus/`aria-activedescendant`.
4. `aria-checked` on each option to explicitly communicate applied state.
5. An apply button.

I suspect that in the above, point `#3` is what people are going to have trouble with.  Why not just use `aria-selected` to indicate applied state?  These aren't multiselect controls, after all.

As an example of why that is problematic, screen readers have inconsistent support for conveying selected versus not selected state.  Both in terms of when they do and don't convey it, and in the speech patterns that they use.  I mention both of these points fully aware that exact assistive technology output is somewhat outside of the scope of the APG, but also because it supports the spec being explicit about support for various attributes in mark-up.  It has also been my experience that users understand checked versus not checked a lot more than selected or not selected versus screen reader silence (for example because on something like the Windows desktop, select state isn't usually spoken/brailled).

Note: point `#4` above is why I wouldn't disagree with some explicit new roles of "optioncheckbox" or "treeitemcheckbox".

**Comment by scottaohara** [comment]
@craigkovatch i would submit the the material component you linked to is not at all what's being discussed here.  As far as that not being a persuasive argument, fine?  I mean, pointing to bad ux to say "see it's ok!" doesn't make it ok.  "authors going to author" yes.  but i highly disagree that a spec is the place to introduce bad practices as if they aren't going to have implications somewhere else (html, wcag, etc.).

**Comment by craigkovatch** [comment]
@scottaohara I see a listbox with items which have a secondary, interactive child, which are sometimes checkboxes and sometimes buttons. So relevant to both "items which can be checked and/or selected" discussion and also @joanmarie's suggestion "then we probably should make options not children-presentational so authors can add checkable descendants", i.e. your immediately previous reply in point number 3...?

**Comment by scottaohara** [comment]
It's not a listbox.  Those are separate controls taking up space on the same horizontal plane.  Just doing basic keyboard tabbing you will notice there are two stops (one unfortunately on the parent wrapping div that then targets down to the visually hidden checkbox input, and the second on the button).  Looking at the DOM, while they are nested within a clickable nonsense div, they are not invalidly nested in a way that would be in misalignment to HTML or ARIA right now.

**Comment by smhigley** [comment]
I want to bring up again that no one has been able to find another real-world example of dual state yet. I do want to establish whether there's actually a real-world use case for this, because otherwise there should be a pretty simple path forward. Dual on/off state (persistent state, not focus) is not a thing I've encountered (again despite working with hundreds of apps), which is part of what makes this discussion so weird to me. I also reached out to other UX folks and control pattern developers to ask if they're familiar with it, and they are not. In fact, the common reaction was along the lines of "WTF, how does that work?".

@jscholes the point about preferring checked state announcements is well made, though I think for your third example you could easily leave off `aria-selected` entirely, and just use `aria-checked` + focus/activedescendant. It still sounds like an example with only one on/off state (checked) in addition to the point of regard.

I 100% agree with @joanmarie  and @sinabahram that "Authors gonna author," and we should support accessibility, even if not perfect, for common and emergent patterns that address actual UI needs. The thing is, I haven't seen any evidence on this thread that dual independent selected/checked states is actually a thing that authors need or even create in practice. Earlier I was trying to demonstrate why that is the case (i.e. all the accessibility and usability problems it introduces). Maybe I should be more direct: do we have any examples to show that allowing separate dual state a) addresses a user need, or b) is a common pattern?

So far we only have one very flawed real world example of this from Tableau, which [on further investigation](https://public.tableau.com/profile/tableau.for.sales.analytics#!/vizhome/SalesPipelineDashbaord/PipelineDash) doesn't even use mixed `aria-selected`/`aria-checked`. Can anyone find other examples to support this being something people do? If we can't find those, we could continue to support both `aria-checked` and `aria-selected` on option and treeitem in the spec, but specify that they cannot both be defined for the same control. 

Also, because the question about WinForms and other Windows selection/checking patterns has been raised before, I did some research on it for one of the past deep dives. I'll paste the specifics below, but the TL:DR; is that while Windows has used both checked and selected states in the past, it has not enabled independent control of them as two separate on/off states on the same option. In more detail:
- Winforms: the "Checked List" control implements both checked and selected semantics, but you cannot separately control checked and selected state
- WPF: The selection list has visual checkboxes, but only implements selection under the hood
- UWP XAML: Does not have visual checkboxes, and only implements selection under the hood

**Comment by sinabahram** [comment]
@smhigley In the spirit of providing examples, may I please ask for an example of a treeview that allows for multiple selection and checked state in the way that you think is best/ideal e.g. with buttons or however else you feel addresses all of the concerns you have brought up, but still allows for multiple items to be acted upon at once in terms of their states being toggled?

**Comment by jscholes** [comment]
@smhigley

> I think for your third example you could easily leave off aria-selected entirely, and just use aria-checked + focus/activedescendant.

I don't disagree in theory.  I don't believe assistive technology support for such a pattern to be very good (although I would have to carry out exhaustive testing).  The main issue is that at present, I would be going directly against the v1.2 Combo Box and List Box patterns in the APG.

For [Combo Box](https://www.w3.org/TR/wai-aria-practices-1.2/#wai-aria-roles-states-and-properties-6):

> For a combobox that controls a listbox, grid, or tree popup, when a suggested value is visually indicated as the currently selected value, the option, gridcell, row, or treeitem containing that value has aria-selected set to true.

For [List Box](https://www.w3.org/TR/wai-aria-practices-1.2/#listbox_roles_states_props):

> In a single-select listbox, the selected option has aria-selected set to true.

It would certainly be a concern if there was no visual indication of which box a user would be checking and not checking, so `aria-selected` would still be needed in addition to `aria-checked`.  If web authors must actively go against the APG to create the sort of pattern I've described, then that would be an argument for clarifying the attribute usage which seems to be the aim of this thread.

**Comment by smhigley** [comment]
@sinabahram I also have never come across a tree that has dual checked/selected state, so no, I don't have examples of that :). The basic problem I keep trying to raise is that having two separately controlled on/off states on the same item is not usable or understandable on a basic level.

@jscholes if you're using `aria-checked` instead of `aria-selected` it would already be a break from the APG examples, which seems completely fine if you've found an approach you think is better. Adding `aria-selected` to the currently focused option still would not be following the quoted description in the APG, since it would not be applied to the selected option, it would be applied to the current point of regard.

As far as AT support goes, I have not found any difference in support for focus/`activedescendant` when `aria-selected` is absent, nor have any effect on support for `aria-checked`. If anything, having `aria-selected` in addition to `aria-checked` is actively worse for that pattern, because you get confusing announcements like "selected, not checked." If you want to have checked semantics for a multiselect, the best implementation would likely be to leave off `aria-selected` entirely.

**Comment by craigkovatch** [comment]
> So far we only have one very flawed real world example of this from Tableau, which on further investigation doesn't even use mixed aria-selected/aria-checked.

@smhigley You are right that the current public code is (even worse than previously described) a `listbox` container with `checkbox` items. This was a historical mistake of chasing screen reader behavior, rather than implementing a proper spec, and it's on me to fix it. My plan of action is as-described above. This is why I keep asking: if option+checked+selected is not the best way to expose this to AT, what is? (I have separate work pending on me to implement a new multi-select listbox which needs to also work inside of a dropdown, so this discussion is triple-relevant to me right now.)

> I want to bring up again that no one has been able to find another real-world example of dual state yet. I do want to establish whether there's actually a real-world use case for this, [...] we only have one very flawed real world example of this from Tableau
> having two separately controlled on/off states on the same item is not usable or understandable on a basic level.

This is very frustrating. I have provided a real-world use case. I'm struggling to understand why you're dismissing this as a case that somehow doesn't count. Hundreds of thousands of people use this control every day. They apparently use it and understand it just fine. It's admittedly a weird, quirky pattern. I concede again that it is not best practice, and I'm not asking anyone to like it or approve of it. But I don't understand why it being "flawed" is disqualifies it as an "actual" case for this discussion.

I'm trying to do the best thing by our users, and thus trying to provide the feedback on this Issue that at least one real-world example *does* want to use this. I'm frustrated that it has several times now been suggested that onus is somehow on me to find **more** examples for code that neither I nor my company owns. I'm frustrated at your suggestion that ours is not "an actual real-world use case". Why?

**Comment by craigkovatch** [comment]
@scottaohara I am really surprised to hear you say that Material example is "not a listbox". What is it? I'm asking about its conceptual function, not about it's current markup. I don't see how it could be described as anything but a multiselectable listbox with interactive children (which is why I first posted it, since @joanmarie had raised the question of whether allowing those in `option` should be reconsidered).

**Comment by stes-acc** [comment]
Much of the confusion comes from the facht that we are talking about different things here.
Items could be flagged=checked as a result of a previous operation on the dataset. In a subsequent step, some of them are now selected for further operation.

I agree that the "checked" state should not be used for such  "preselection", we would need a custom new attribute for that or indicate such a flagged state using aria-description which would be way better.

Accepting that would render the entire discussion above obsolete IMHO.

**Comment by scottaohara** [comment]
@craigkovatch
>I am really surprised to hear you say that Material example is “not a listbox”.
>I don't see how it could be described as anything but a multiselectable listbox with interactive children...

Why?  I'm not going to call it what it isn't.  It’s a list (`<ul>`) of checkboxes and buttons. This is a pretty common pattern that doesn't touch on what we've been talking about. “Conceptual function” isn't applicable because then we're using a real example with different functionality to validate an argument for something it isn't.  Two separate controls, no keyboard behavior similar to listbox > options, exposed to assistive technologies differently.  Apples and pumpkins comparison.  

If you want to go the route of saying they are visually similar looking... sure? But that can be said for many types of custom controls and content on the web. How it behaves and is exposed, e.g., an example of a selected and checked state for an option in the wild, is what matters here if we're to stay on topic. So, this example doesn't belong here. 

And no, I mean “nonsense div” because it’s a `div` with a `tabindex=0`, no role, no accessible name (not directly at least, one will be cobbled together from its children), and a click event on it that triggers the nested checkbox which has been set to `tabindex=-1`.  Screen readers/browser pairings all announce this pattern differently due to the current setup / them attempting to mitigate the awkward markup (some relatively fine, some with duplicate announcements of the content within, depending on your navigation method).

**Comment by jscholes** [comment]
> In my opinion, the "checked" state should not be used for such "preselection", we would need a custom new attribute for that or indicate such a flagged state using aria-description which would be way better.

I disagree on both accounts.  A new attribute would need to either have custom support in assistive technologies or map to an existing semantic in accessibility APIs across platforms.  Not an easy or guaranteed feet.  `aria-description` and `aria-describedby` are also not intended for surfacing state which should be semantically conveyed, and would introduce all sorts of problems like the inconsistent announcement of an accessible description by screen readers, a user's ability to turn it off, etc.

Regardless, I'm curious what you propose using as either a new attribute or text for `aria-description`?  What could we tell usres that isn't already perfectly conveyed by checked/not checked, states users are completely familiar with?

**Comment by stes-acc** [comment]
> > In my opinion, the "checked" state should not be used for such "preselection", we would need a custom new attribute for that or indicate such a flagged state using aria-description which would be way better.
> 
> I disagree on both accounts. A new attribute would need to either have custom support in assistive technologies or map to an existing semantic in accessibility APIs across platforms. Not an easy or guaranteed feet. `aria-description` and `aria-describedby` are also not intended for surfacing state which should be semantically conveyed, and would introduce all sorts of problems like the inconsistent announcement of an accessible description by screen readers, a user's ability to turn it off, etc.
> 
> Regardless, I'm curious what you propose using as either a new attribute or text for `aria-description`? What could we tell usres that isn't already perfectly conveyed by checked/not checked, states users are completely familiar with?

Misunderstanding. We still have selected state but preselection means flagging with another info. The actual value for aria-description flag depends on that, it could mean "shipped to customer". These custom attributes are VERY common in business apps. But I won't use a checkbox in this case for visuals, icons will do better here.

**Comment by jscholes** [comment]
> it could mean "shipped to customer"

I understand what you're getting at, and `aria-checked` or `aria-selected` would absolutely not imply "shipped to customer".  If either of those was ever used to indicate such state, it would be a pretty confusing UX.  But I don't think that negates the usefulness of `aria-checked` and `aria-selected` in other cases which aren't highly domain-specific.

**Comment by carmacleod** [comment]
I've been trying to think of an example that would use both `aria-checked` and `aria-selected`.
Perhaps this one will qualify.

Please note that:
- this example only exists as a thought experiment (I am not aware of any actual implementation)
- multiselect mouse/keyboard interaction would be intended for power users
- it relies heavily on context menu and/or app-specific shortcut keys and/or action/toolbar buttons (i.e. buttons/toolbar external to the list)

Imagine a To Do list.
- "checked" on an item means "done"
- mouse/keyboard user can select multiple items in the usual ways
- mobile/touch user has an "Edit" button that puts the list in a mode that allows them to select multiple (i.e. tap selects/deselects)
  - a set of action buttons (or a toolbar) is available in edit mode
  - user cannot check/uncheck with tap when in edit mode, but they can use use "mark done" and "mark not done" buttons
  - when not in edit mode, tap checks/unchecks items (i.e. does not select/deselect)
- once multiple items have been selected,
  - mouse/keyboard users can use context menu or shortcut keys to choose an action
  - all users can use action buttons/toolbar to choose an action
  - possible actions might be "move up", "move down", "priority", "due date", "remove" (which can have delete key shortcut)
- whether or not to support space key to toggle "checked" state when there are multiple items selected would be up to the design
  - might want to only support if one item is selected, or if multiple items with the same checked state are selected
  - if space is not supported (probably even if it is), then actions for "mark done" and "mark not done" would be in the context menu/shortcuts/toolbar

Another example (same interaction, different use case) might be a list of breakpoints, where "checked" means "enabled".
User can select multiple to "enable", "disable", "delete", "export", etc.

In these thought experiments,
- the "checked" state is permanent, likely saved somewhere in the app. It takes an explicit user action to change it.
- the "selected" state is transient, and is only used to indicate the set of items that will participate in an action. The "selected" state changes as soon as the user changes their point of regard.

I'm not saying this is necessarily a good design. The app would want to make it clear, either through documentation, labelling or describing, that "checked === done" and "selected === ready for action".

Space key requires extra design thinking because if the user selects some checked and some unchecked items and hits space, then what? Error message? Toggle anyhow? Always check? Do nothing? I'll let the design team decide what makes sense for the app. :)

**Comment by mcking65** [comment]
@jscholes  wrote:

> As a practical example, the sidebar on this GitHub issue contains several "dropdowns" (for lack of a better term) for applying assignees, labels, etc.

James, That input in the side bar shouldn't be a combobox because when you choose an item from the list it is not setting the value of the input. A combobox is an input that has a popup that helps set the value of the input. Choosing items in the popup should close the popup and put that value into the input.

The edit field that appears with the label chooser in GitHub is simply a filter. It filters what appears in the list of tags. That list is basically a multi-select listbox where selection means that tag has been applied to the issue.

Since the tags chooser widget is a multi-select, selection should not follow focus. Instead, the user moves focus to a tag and should then press space to toggle the selected state. In that implementation, it is also perfectly valid to use aria-checked instead of aria-selected to represent the state of the tag. There is no need for both. Personally, in multi-selects like that, I prefer aria-checked. For some reason it seems more clear to me. But, I probably only have that bias because of my background. I would strongly suggest using aria-checked if the state is visually represented with a check mark.

I believe the most appropriate ARIA patterns for the GitHub label chooser are a disclosure button that shows/hides the filter textbox and a multi-select listbox. Three very simple patterns that are all well-supported -- nothing tricky.

At this time, it is using a menubutton and a menu ... and it does not work well. That combination could possibly work without the filter. But, ARIA does not support textboxes inside of menus.

**Comment by jscholes** [comment]
@mcking65 

> That input in the side bar shouldn't be a combobox ... That list is basically a multi-select listbox

Agreed on these semantics.  But they would still need to use `aria-checked` on `role="option"` in a list box implementation.

> In that implementation, it is also perfectly valid to use aria-checked instead of aria-selected to represent the state of the tag. There is no need for both. Personally, in multi-selects like that, I prefer aria-checked.

We also agree on this.  But I don't believe the support for `aria-checked` on `role="option"` to be in a good state across assistive technology.  That is, of course, not the responsibility of the APG Task Force, but it's also not surprising given the focus on `aria-selected` in virtually all discussions of list boxes.

For example, with NVDA/Chrome, a checked state is announced but not an unchecked one, meaning that users have no idea they can toggle the current item.  On MacOS with VoiceOver, neither checked nor not checked are announced.  So while the pattern may be valid, it would be irresponsible to actually suggest that anybody implement it.

Again, I'm mindful of the title of this issue, `aria-checked on option and treeitem`.  In the first post of the thread, @smhigley asserted the following:

> I believe it should be removed from option (and treeitem)

(where "it" refers to `aria-checked`).  If having `aria-checked` on both of these roles is valid, and a pattern that should be recommended to web authors once assistive technology support is improved, then the actual matter under discussion is whether `aria-selected` and `aria-checked` should be allowed in tandem, not whether `aria-checked` should be removed from `option` and `treeitem`.  You can forgive AT developers and web authors not having consensus when this issue exists suggesting it should be removed entirely from these roles.

**Comment by craigkovatch** [comment]
@jscholes 

> For example, with NVDA/Chrome, a checked state is announced but not an unchecked one, meaning that users have no idea they can toggle the current item. On MacOS with VoiceOver, neither checked nor not checked are announced. So while the pattern may be valid, it would be irresponsible to actually suggest that anybody implement it.

That sounds like some bugs needs to be filed on screen readers, then :)

**Comment by jscholes** [comment]
@craigkovatch 

> That sounds like some bugs needs to be filed on screen readers, then

I agree.  But the aim of this thread seems to be to make a determination about whether `aria-checked` is valid on `role="option"` and `role="treeitem"`.  It would help in making a case to AT developers if consensus was reached here, given that the thread is proposing disallowing it altogether.

To be clear, I have an interest in `aria-checked` being consistently supported by AT.  I agree with the assertions that having dual-toggleable checked and selected states on something could be confusing, although can see a place for it in a specialised UI.  But I also feel that the core question being tackled here as been lost and overrided by, should we allow `aria-selected` and `aria-checked`, together?  I don't think the answer to that question is to disallow `aria-checked` on `option` and `treeitem`.

@smhigley What are your thoughts?  In the first post of this thread, you wrote:

> updating to a more strict set of allowed attributes seems beneficial

But in response to my GitHub sidebar example, you wrote:

> you could easily leave off aria-selected entirely, and just use aria-checked + focus/activedescendant.

These two don't necessarily work together, if you're proposing that `aria-checked` be disallowed on `option` and `treeitem`.  So the answer seems to be, allow `aria-selected` and `aria-checked`, but in a mutually exclusive relationship.

**Comment by smhigley** [comment]
For use during the call today, I put together a working example of each of the patterns we've been discussing: https://quirky-meitner-4a68e8.netlify.app/. I added an `alert()` to simulate the idea of something "happening" on check/uncheck for the first one, and on confirmation for the second.

@jscholes I know the original proposal was to disallow `aria-checked` on `option` (I think `treeitem` has been pretty thoroughly dropped by now), but I think it would also make sense to specify that checked and selected can't co-exist (what Leonie suggested much earlier: https://github.com/w3c/aria/issues/1052#issuecomment-536018316). That would solve the issue where devs confuse selected with point of regard, and accidentally implement both :).

**Comment by cookiecrook** [comment]
I mouse clicked the first checkbox, then after clearing the JS alert, shift+clicked the fifth, resulting in this:

![screen shotfirst and fifth options checked, and first through fifth selected](https://user-images.githubusercontent.com/75867/94181771-d0158280-fe54-11ea-8607-f187098c7a56.png)

I'm not sure why there is a both a multi-selection state and a multi-checked state.

**Comment by jscholes** [comment]
@smhigley Thanks for clarifying your position and the samples.  Will this URL remain active so that it can be used in AT/browser bug reports if necessary?  Or can it be forked?

Everyone will have a different dog in this fight based on what they need to support, or would like/participate needing to support in the future.  I've made it clear that my priority is consistent support for `aria-checked` on `option` and `treeitem`, because I find it to be a useful paradigm in many UI patterns, and have encountered a number of client requirements which couldn't be properly semantically supported because of the current lack of consistency.  I am satisfied that many other people here seem to feel the same.

Having said that, it is my strong opinion that web authors should not be constrained based on what the ARIA Working Group believes to be a usable or pertinent pattern.  We've all been around the block enough times to know that if a web or native control doesn't provide exactly what a web author or other stakeholder feels is required, somebody will invent their own.  ARIA exists to ensure that, good or bad ideas aside, users of assistive technologies can at least be supported or taken into account.  If the users of that pattern find it to be frustrating, inappropriate or even blocking, it is for them to raise that with their employer or other applicable entity, and not for the ARIA Working Group to dictate what should be done.  That's my understanding, anyway; happy to be corrected!

Anecdotally, I would happily implement a dual-state list box or tree view if I could guarantee accessibility or the availability of training for the target audience, particularly if one of those states was a convenience factor and not strictly necessary to complete a task.  For instance, inside an accessibility test reporting tool, I may want to claim that an issue fails multiple contiguous criteria.  The checked state would determine whether a given criterion was applicable, but it would also be helpful to be able to press Shift+Down Arrow a few times and check all of the selected criteria in one go.  Necessary?  No.  But certainly helpful.

I understand that there are concerns with a UI like the one outlined above.  It may be extremely difficult to make accessible for speech recognition users, and surfacing the dual state visually without relying on colour alone may be a challenge.  But real-world constraints mean that somebody may implement such an UX regardless, and we shouldn't deny support for their endeavours purely from an ARIA perspective.  When we work with them on a voluntary or professional basis as accessibility practitioners, or  provide them with feedback as customers, that's when fundamental design concerns can be raised.

**Comment by stes-acc** [comment]
There was a question why we have selected AND checked states for treeitems.

I remember having "checked" on treeitem to reflect **mixed** states in MS installation trees like below

[*] All
-  [X] Item 1
-  [  ] Item 1
-  [X] Item 1

You can't map that properly with selected state. And selected was in because every tree item is selected implicitly when focused as list definition says.

**Comment by jscholes** [comment]
@stes-acc I think GitHub's Markdown renderer interfered with what you were trying to convey there, but an installer is certainly a great example of where `aria-checked` on treeitems and options is useful and familiar to users.  I don't know how `aria-selected` would enter into the equation; there is some related discussion over on https://github.com/w3c/aria/issues/1311 where at @carmacleod correctly pointed out:

> NVDA verbosely says "not selected" for every treeitem with `aria-selected="false"`

If this is also the case for items with no `aria-selected` application whatsoever, it would be an argument for `aria-selected` to follow focus inside an installer, with `aria-checked` being used to opt in/out of specific items.

**Comment by mcking65** [comment]
@jscholes wrote:

> If this is also the case for items with no `aria-selected` application whatsoever, it would be an argument for `aria-selected` to follow focus inside an installer, with `aria-checked` being used to opt in/out of specific items.

If aria-checked is used to indicate the opt in/out state in the installer, why would selection follow focus? Why should there be a selected state at all?

**Comment by jscholes** [comment]
@mcking65

> why would selection follow focus? Why should there be a selected state at all?

I agree that there shouldn't need to be.  My comments were specifically in line with those over at the related issue (#1311) where it was pointed out that there is currently confusion about when `aria-selected` is and is not required, which is in turn causing assistive technology support to be inconsistent.  And in NVDA's case, overly verbose to the point where I could not, in good faith, tell somebody to only use `aria-checked` in a tree.

**Comment by jscholes** [comment]
Also note: the verbosity of NVDA was explicitly observed for treeitems with `aria-selected="false"`.  I noted in my comment:

> If this is also the case for items with no aria-selected application whatsoever

In other words, if NVDA announces "not selected" for every treeitem with no `aria-selected` attribute, assuming the value of a missing attribute to be `"false"`, selection would need to follow focus right now to make for a pleasing speech and braille experience.  This may not be the case and would need to be tested.

**Comment by smhigley** [comment]
@jscholes wrote:
> Will this URL remain active so that it can be used in AT/browser bug reports if necessary? Or can it be forked?

I'll probably be experimenting a bit with that URL, so it's not the best to use to file bugs. If you want, the source code is here: https://github.com/smhigley/aria-mocks. You don't really need a working example like this to file bug reports, though; just some static HTML with the roles/attributes should be plenty.

The issue here is still the conflict between checked and selected. Asking authors to always set selected to the currently focused item is not a good spec solution as @mcking65 mentioned, since selection is conceptually not the same thing as focus. You're also not consistently reducing verbosity by setting it, since multiple screen readers will read the selected state in addition to not selected. VoiceOver will even only read "selected" but not "not selected," so there isn't really even a good practical workaround there.

re the ARIA spec adding constraints to authors: there are plenty of constraints already in the spec. Most of the characteristics table directly deals with adding constraints; `aria-checked` is not supported on tab items, for example. This isn't really about opinions on whether the pattern is subjectively good or bad, the issue here is that logistically, especially for options, checked is poorly defined, inconsistent, and poorly supported. It's also not just a matter of pushing browsers and screen readers to add support; the desired behavior is not even defined in the spec.

**Comment by mcking65** [comment]
@smhigley, @aleventhal, and I have been meeting offline to work on proposals that we hope will lead to consensus.

I have started to draft some language in PR #1340. Keep in mind, this is draft that is inspired by on-going conversation; it does not represent conclusions from that conversation.

In the spirit of looking for middle ground that avoids placing constraints on authors that might not be essential, here is draft authoring guidance for the option role that is included in the pull request.

>The option role supports both aria-selected and aria-checked because it is common for authors to use selection in single-select list boxes and check marks in multi-select list boxes. However, authors SHOULD NOT specify both aria-selected and aria-checked on option elements contained by the same listbox except in the extremely rare circumstances where all the following conditions are met:
>
>* The meaning and purpose of aria-selected  is different from the meaning and purpose of aria-checked  in the user interface. 
>* The user interface makes the mening and purpose of each state apparent. 
>* The user interface provides a separate method for controling each state. 

I include a bit more color on the rationale as well as explanation of other changes in my comment in #1340.

**Comment by mcking65** [comment]
I think there may be an assumption underneath some comments in this thread that I'd like to bring to the surface in the hope that clarity about assumptions can help inspire understanding of differing perspectives.

I think the gist of the assumption is something like the following.

>The ARIA specification should only allow for semantics that work for all users in any user interface. Or conversely, if some semantics could create problems for some users in some circumstances, then they should not be supported by the specification. Example: it is not advisable for the specification to support semantics that would be useful to users who can use a keyboard but impractical for users who rely on touch. Or, it is not advisable to support semantics that could create a user interface that would be problematic for users with cognitive disabilities.

I think the spirit of the assumption is good. But, I'm hesitant to endorse the ARIA specification as a tool for limiting authors abilities to serve a specific audience with a specific solution or even to enforce good design principles.

To use a building construction analogy, I've thought of the scope of ARIA to be most comparable to the specifications for building materials, e.g., it specifies the dimensions and load baring capacity of a specific truss. ARIA does not tell you how far apart you should space your trusses in a specific circumstance. For that, we turn to the building code -- WCAG. It is also important to recognize the limits of WCAG. If you were to design a home using a building code as your definition of good design, it would be unsurprising if the home were unliveable. For instance, you might find the toilet positioned between the kitchen stove and sink, and from what I know about private residential codes, most would permit it as long as spacing requirements are met.

**Comment by sinabahram** [comment]
I just want to express sincere gratitude to @smhigley, @mcking65, and @aleventhal for your work on this RE the above that Matt posted as a working summary. Also Matt, thank you for the clarification around purpose as that matches my original understanding of the ARIA specification.

**Comment by craigkovatch** [comment]
@mcking65 really appreciate the effort, commentary, and summaries here. Thank you and everyone else who contributed.

Would like to ask a clarifying question via scenario: say I have a combobox, with a multi-select listbox popup. DOM focus of course remains in the `<input>`, while "item focus" (AT focus? visual focus? what's the term of art here?) is indicated using `aria-activedescendant`. I have seen past posts from @smhigley which indicated that VoiceOver doesn't read `aria-activedescendant` updates unless `aria-selected` is also updated on the focused `option` (not sure if this is still current, I think I remember reading that ca. 2018).

This is a scenario in which I would expect to use `aria-checked` (against my semantic instinct) to indicate "membership in the set of options which the user has selected", while using `aria-selected` + `aria-activedescendant` for indicating the "active" item for keyboard interaction. Does this track? Is there a better way? I think this still works within the draft constraints laid out, but seems like a useful test scenario given how complex comboboxes always are.

(Looks like this is exactly what SalesForce does in Lightning comboboxes, FWIW: https://www.lightningdesignsystem.com/components/combobox/#Multiple-selections)

**Comment by mcking65** [comment]
@craigkovatch  wrote:

>I have seen past posts from @smhigley which indicated that VoiceOver doesn't read `aria-activedescendant` updates unless `aria-selected` is also updated on the focused `option`.

This would be a VoiceOver bug. If it is still present, we'll catch it in the aria-at project.

However,  this bug would only negatively impact a free-standing listbox. If a listbox is a combobox popup, selection follows focus so that bug would be inconsequential.

> This is a scenario in which I would expect to use `aria-checked` (against my semantic instinct) to indicate "membership in the set of options which the user has selected", while using `aria-selected` + `aria-activedescendant` for indicating the "active" item for keyboard interaction. Does this track?

As of ARIA 1.2, combobox is a single-select widget. The value of the combobox is one item chosen from the popup.

If a future version of combobox were to support choosing multiple items for the value of the combobox, you would still only use either checked or selected, not both. The VoiceOver bug would have to be fixed if VoiceOver were to support a multi-select combobox.

>I think this still works within the draft constraints laid out, but seems like a useful case for the testing the assumptions of whatever gets drafted.

The constraints in the PR would definitely permit a future version of combobox that supports multi-select. But, again, as described in the PR, you would use only checked or selected, not both.

As of today, it is not a useful test case because of current combobox constraints.

**Comment by craigkovatch** [comment]
@mcking65 
> As of ARIA 1.2, combobox is a single-select widget. The value of the combobox is one item chosen from the popup.

Reviewing https://www.w3.org/TR/wai-aria-1.2/#combobox, I don't see anything that speaks to this. Is this not yet reflected in the current Working Draft?

**Comment by JAWS-test** [comment]
I also think that the ARIA specification for combobox does not currently prohibit multiple selection. In the [APG](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox) it becomes clearer that only one value can be selected. I.e. the specification should say here more clearly that multiple selection is not allowed. 

In the future, multiple selection should be possible with combobox and I am in favor of implementing this as soon as possible. Many frameworks have been offering comboboxes with multiple selection for a long time now, but they all don't work well because the corresponding ARIA support is missing
